### PR TITLE
[ECO-574] Arvato - AfterPay

### PR DESCRIPTION
**Pull request by IvanovRuslanSpryker**
- Developer(s): @IvanovRuslanSpryker

- Ticket: https://spryker.atlassian.net/browse/ECO-574

- Documentation: https://spryker.atlassian.net/wiki/spaces/ECO/pages/788071193/Afterpay+Documentation

#### Please confirm

- [ ] No new OS components - or they have been approved by legal department

#### Documentation

- [ ] Functional documentation provided or in progress?
- [ ] Integration guide for projects provided or not needed?
- [ ] Migration guides for all contained majors provided or not needed?

#### Release Table

   Module                | Release Type         | Constraint Updates         |
   :--------------------- | :------------------------ | :--------------------- |
   AfterPay              | major (new)           |                       |

#### Release Notes

Added integration functionality for AfterPay.

-----------------------------------------

##### Change log

Initial Release

* Next payment methods have been added:
-- Invoice;
* Added two checkout process:
-- One-Step Authorization → AFTERPAY_AUTHORIZE_WORKFLOW_ONE_STEP;
-- Two-Step Authorization → AFTERPAY_AUTHORIZE_WORKFLOW_TWO_STEPS;
* Added functional tests.

### CODE REVIEW

**Review by alex-galych — APPROVED** [review]
Reviewed in internal reviews.

File: .travis.yml [hunk 1]
```diff
@@ -43,5 +43,5 @@ before_install:
   - echo "extension = redis.so" >> ~/.phpenv/versions/$(phpenv version-name)/etc/php.ini
 
 script:
-  - git clone -b 0.2.0 https://github.com/spryker-eco/eco-ci.git ecoci
-  - ./ecoci/build/travis.sh
+  - git clone -b 0.2.1 https://github.com/spryker-eco/eco-ci.git ecoci
+  - ./ecoci/build/travis.sh
```
Comment (lockw1n) [inline .travis.yml]: Please add new line at the of the file.

File: phpstan.neon [hunk 2]
```diff
@@ -1,9 +1,6 @@
 parameters:
     bootstrap: %rootDir%/../../../phpstan-bootstrap.php
 
-    ignoreErrors:
-    - '#Call to an undefined method .+LocatorLocatorInterface::.+\(\).#'
-
 services:
```
Comment (lockw1n) [inline phpstan.neon:4]: This section has to be removed with all its items.

File: src/SprykerEco/Yves/AfterPay/Form/DataProvider/InvoiceDataProvider.php [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+class InvoiceDataProvider implements StepEngineFormDataProviderInterface
+{
+    /**
+     * @param \Generated\Shared\Transfer\QuoteTransfer $quoteTransfer
+     *
+     * @return \Generated\Shared\Transfer\QuoteTransfer
+     */
+    public function getData(AbstractTransfer $quoteTransfer): QuoteTransfer
+    {
+        if ($quoteTransfer->getPayment() === null) {
```
Comment (lockw1n) [inline src/SprykerEco/Yves/AfterPay/Form/DataProvider/InvoiceDataProvider.php]: We prefer early return.

File: src/SprykerEco/Zed/AfterPay/AfterPayConfig.php [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+    }
+
+    /**
+     * @param string $paymentMethod
+     *
+     * @return string
+     */
+    public function getPaymentChannelId(string $paymentMethod): string
+    {
+        switch ($paymentMethod) {
```
Comment (lockw1n) [inline src/SprykerEco/Zed/AfterPay/AfterPayConfig.php]: This switch does not have any sense. And also business logic should not be present in config class.

File: src/SprykerEco/Zed/AfterPay/Business/AdditionalServices/Handler/LookupCustomerHandler.php [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+ * For full license information, please view the LICENSE file that was distributed with this source code.
+ */
+
+namespace SprykerEco\Zed\AfterPay\Business\AdditionalServices\Handler;
+
+use Generated\Shared\Transfer\AfterPayCustomerLookupRequestTransfer;
+use Generated\Shared\Transfer\AfterPayCustomerLookupResponseTransfer;
+use SprykerEco\Zed\AfterPay\Business\Api\Adapter\AdapterInterface;
+
+class LookupCustomerHandler implements LookupCustomerHandlerInterface
```
Comment (lockw1n) [inline src/SprykerEco/Zed/AfterPay/Business/AdditionalServices/Handler/LookupCustomerHandler.php:14]: In general we have accepted RFC that we disallow handlers. But you don't need to fix it right now. You can create a task into technical debts and keep in mind in further integrations.

File: src/SprykerEco/Zed/AfterPay/Business/AfterPayBusinessFactory.php [hunk 6]
```diff
@@ -0,0 +1,508 @@
...
+use SprykerEco\Zed\AfterPay\Business\Payment\Transaction\TransactionLogReader;
+use SprykerEco\Zed\AfterPay\Business\Payment\Transaction\TransactionLogReaderInterface;
+use SprykerEco\Zed\AfterPay\Dependency\Facade\AfterPayToCustomerFacadeInterface;
+use SprykerEco\Zed\AfterPay\Dependency\Facade\AfterPayToMoneyFacadeInterface;
+use SprykerEco\Zed\AfterPay\Dependency\Facade\AfterPayToPaymentFacadeInterface;
+use SprykerEco\Zed\AfterPay\Dependency\Facade\AfterPayToStoreFacadeInterface;
+use SprykerEco\Zed\AfterPay\Dependency\Service\AfterPayToUtilEncodingServiceInterface;
+
+/**
+ * @method \SprykerEco\Zed\AfterPay\Persistence\AfterPayQueryContainerInterface getQueryContainer()
```
Comment (lockw1n) [inline src/SprykerEco/Zed/AfterPay/Business/AfterPayBusinessFactory.php:84]: QueryContainers should not be used. Please also put it into technical debts.

File: src/SprykerEco/Zed/AfterPay/Business/Api/Adapter/AfterPayApiAdapter.php [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+
+    /**
+     * @param \Generated\Shared\Transfer\AfterPayAvailablePaymentMethodsRequestTransfer $requestTransfer
+     *
+     * @return \Generated\Shared\Transfer\AfterPayAvailablePaymentMethodsResponseTransfer
+     */
+    public function sendAvailablePaymentMethodsRequest(AfterPayAvailablePaymentMethodsRequestTransfer $requestTransfer): AfterPayAvailablePaymentMethodsResponseTransfer
+    {
+        return $this
+            ->adapterFactory
```
Comment (lockw1n) [inline src/SprykerEco/Zed/AfterPay/Business/Api/Adapter/AfterPayApiAdapter.php]: I would prefer to move it on previous line
```
return $this->adapterFactory
    ->createAvailablePaymentMethodsCall()
    ->execute($requestTransfer);
```

Please change methods in this file.

File: src/SprykerEco/Zed/AfterPay/Business/Api/Adapter/ApiCall/AuthorizePaymentCall.php [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+     *
+     * @return \Generated\Shared\Transfer\AfterPayApiResponseTransfer
+     */
+    protected function buildAuthorizeResponseTransfer(string $jsonResponse): AfterPayApiResponseTransfer
+    {
+        $jsonResponseArray = $this->utilEncoding->decodeJson($jsonResponse, true);
+
+        $responseTransfer = new AfterPayApiResponseTransfer();
+
+        $responseTransfer
```
Comment (lockw1n) [inline src/SprykerEco/Zed/AfterPay/Business/Api/Adapter/ApiCall/AuthorizePaymentCall.php]: You can merge it with transfer creation.
```
$responseTransfer = (new AfterPayApiResponseTransfer())
    ->set()
    ->set();
```

File: src/SprykerEco/Zed/AfterPay/Business/Api/Adapter/ApiCall/AvailablePaymentMethodsCall.php [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+    protected function buildAvailablePaymentMethodsResponseTransfer(string $jsonResponse): AfterPayAvailablePaymentMethodsResponseTransfer
+    {
+        $jsonResponseArray = $this->utilEncoding->decodeJson($jsonResponse, true);
+
+        $responseTransfer = new AfterPayAvailablePaymentMethodsResponseTransfer();
+
+        $riskCheckResultCode = $this->extractRiskCheckCode($jsonResponseArray);
+        $customerNumber = $this->extractCustomerNumber($jsonResponseArray);
+
+        $responseTransfer
```
Comment (lockw1n) [inline src/SprykerEco/Zed/AfterPay/Business/Api/Adapter/ApiCall/AvailablePaymentMethodsCall.php]: Same. Please merge it with transfer creation.

File: src/SprykerEco/Zed/AfterPay/Business/Api/Adapter/ApiCall/LookupCustomerCall.php [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+    /**
+     * @param array $userProfile
+     *
+     * @return \Generated\Shared\Transfer\AfterPayUserProfileTransfer
+     */
+    protected function buildUserProfileTransfer(array $userProfile): AfterPayUserProfileTransfer
+    {
+        $userProfileTransfer = new AfterPayUserProfileTransfer();
+
+        $userProfileTransfer
```
Comment (lockw1n) [inline src/SprykerEco/Zed/AfterPay/Business/Api/Adapter/ApiCall/LookupCustomerCall.php]: Please merge it with transfer creation.

File: src/SprykerEco/Zed/AfterPay/Business/Api/Adapter/ApiCall/LookupCustomerCall.php [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+    /**
+     * @param array $userAddress
+     *
+     * @return \Generated\Shared\Transfer\AfterPayLookupAddressTransfer
+     */
+    protected function buildLookupAddressTransfer(array $userAddress): AfterPayLookupAddressTransfer
+    {
+        $lookupAddressTransfer = new AfterPayLookupAddressTransfer();
+
+        $lookupAddressTransfer
```
Comment (lockw1n) [inline src/SprykerEco/Zed/AfterPay/Business/Api/Adapter/ApiCall/LookupCustomerCall.php]: Same please.

File: src/SprykerEco/Zed/AfterPay/Business/Payment/Mapper/OrderToRequestTransfer.php [hunk 12]
```diff
@@ -0,0 +1,410 @@
...
+
+    /**
+     * @var \SprykerEco\Zed\AfterPay\Dependency\Facade\AfterPayToStoreFacadeInterface
+     */
+    protected $storeFacade;
+
+    /**
+     * @var array
+     */
+    protected static $paymentMethods = [
```
Comment (lockw1n) [inline src/SprykerEco/Zed/AfterPay/Business/Payment/Mapper/OrderToRequestTransfer.php:48]: Would be nice to move it to constant.

File: src/SprykerEco/Zed/AfterPay/Business/Payment/Mapper/QuoteToRequestTransfer.php [hunk 13]
```diff
@@ -0,0 +1,338 @@
...
+     * @param \Generated\Shared\Transfer\QuoteTransfer $quoteTransfer
+     *
+     * @return \Generated\Shared\Transfer\AfterPayRequestCustomerTransfer
+     */
+    protected function buildCustomerRequestTransfer(QuoteTransfer $quoteTransfer): AfterPayRequestCustomerTransfer
+    {
+        $quoteBillingAddressTransfer = $quoteTransfer->getBillingAddress();
+        $customerRequestTransfer = new AfterPayRequestCustomerTransfer();
+
+        $customerRequestTransfer
```
Comment (lockw1n) [inline src/SprykerEco/Zed/AfterPay/Business/Payment/Mapper/QuoteToRequestTransfer.php]: Please merge it with transfer creation.

File: src/SprykerEco/Zed/AfterPay/Business/Payment/Mapper/QuoteToRequestTransfer.php [hunk 14]
```diff
@@ -0,0 +1,338 @@
...
+    /**
+     * @param \Generated\Shared\Transfer\ItemTransfer $itemTransfer
+     *
+     * @return \Generated\Shared\Transfer\AfterPayRequestOrderItemTransfer
+     */
+    protected function buildOrderItemRequestTransfer(ItemTransfer $itemTransfer): AfterPayRequestOrderItemTransfer
+    {
+        $orderItemRequestTransfer = new AfterPayRequestOrderItemTransfer();
+
+        $orderItemRequestTransfer
```
Comment (lockw1n) [inline src/SprykerEco/Zed/AfterPay/Business/Payment/Mapper/QuoteToRequestTransfer.php]: Same. Merge it please.

File: src/SprykerEco/Zed/AfterPay/Business/Payment/Mapper/QuoteToRequestTransfer.php [hunk 15]
```diff
@@ -0,0 +1,338 @@
...
+     * @param \Generated\Shared\Transfer\QuoteTransfer $quoteTransfer
+     *
+     * @return \Generated\Shared\Transfer\AfterPayRequestAddressTransfer
+     */
+    protected function buildCustomerBillingAddressRequestTransfer(QuoteTransfer $quoteTransfer): AfterPayRequestAddressTransfer
+    {
+        $customerAddressTransfer = $quoteTransfer->getBillingAddress();
+        $customerAddressRequestTransfer = new AfterPayRequestAddressTransfer();
+
+        $customerAddressRequestTransfer
```
Comment (lockw1n) [inline src/SprykerEco/Zed/AfterPay/Business/Payment/Mapper/QuoteToRequestTransfer.php]: Same.

File: src/SprykerEco/Zed/AfterPay/Business/Payment/PaymentReader.php [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+    {
+        $this->afterPayQueryContainer = $afterPayQueryContainer;
+    }
+
+    /**
+     * @param int $idSalesOrder
+     *
+     * @return \Generated\Shared\Transfer\AfterPayPaymentTransfer
+     */
+    public function getPaymentByIdSalesOrder(int $idSalesOrder): AfterPayPaymentTransfer
```
Comment (lockw1n) [inline src/SprykerEco/Zed/AfterPay/Business/Payment/PaymentReader.php:36]: The name of methods like this should be `findPayment...()`. Put it to technical debts.

File: src/SprykerEco/Zed/AfterPay/Business/Payment/PaymentWriter.php [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+        $this->afterPayQueryContainer = $afterPayQueryContainer;
+    }
+
+    /**
+     * @param string $idReservation
+     * @param int $idSalesOrder
+     *
+     * @return void
+     */
+    public function setIdReservationByIdSalesOrder(string $idReservation, int $idSalesOrder): void
```
Comment (lockw1n) [inline src/SprykerEco/Zed/AfterPay/Business/Payment/PaymentWriter.php:35]: The name of methods like this should be `savePayment...()`. Put it to technical debts.

File: src/SprykerEco/Zed/AfterPay/Business/Payment/PaymentWriter.php [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+    {
+        $afterPayPaymentEntity = $this->getPaymentEntityByIdSalesOrder($idSalesOrder);
+        $afterPayPaymentEntity
+            ->setIdReservation($idReservation)
+            ->save();
```
Comment (lockw1n) [inline src/SprykerEco/Zed/AfterPay/Business/Payment/PaymentWriter.php:40]: This should be through entityManager. Put it to technical debts.

File: src/SprykerEco/Zed/AfterPay/Communication/Plugin/Oms/Command/AuthorizePlugin.php [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+use Spryker\Zed\Oms\Business\Util\ReadOnlyArrayObject;
+use Spryker\Zed\Oms\Dependency\Plugin\Command\CommandByOrderInterface;
+
+/**
+ * @method \SprykerEco\Zed\AfterPay\Business\AfterPayFacadeInterface getFacade()
+ * @method \SprykerEco\Zed\AfterPay\Communication\AfterPayCommunicationFactory getFactory()
+ * @method \SprykerEco\Zed\AfterPay\Persistence\AfterPayQueryContainer getQueryContainer()
+ * @method \SprykerEco\Zed\AfterPay\AfterPayConfig getConfig()
+ */
+class AuthorizePlugin extends AbstractPlugin implements CommandByOrderInterface
```
Comment (lockw1n) [inline src/SprykerEco/Zed/AfterPay/Communication/Plugin/Oms/Command/AuthorizePlugin.php:23]: In general we do not allow logic in plugins. Pease put it into technical debts.

**Review by lockw1n — CHANGES_REQUESTED** [review]

File: src/SprykerEco/Zed/AfterPay/Business/Payment/Mapper/OrderToRequestTransfer.php (hunk 12 above)
Comment (IvanovRuslanSpryker) [reply src/SprykerEco/Zed/AfterPay/Business/Payment/Mapper/OrderToRequestTransfer.php:48]: I'm not sure.

File: phpstan.neon (hunk 2 above)
Comment (IvanovRuslanSpryker) [reply phpstan.neon:4]: Please clarify.

**Review by lockw1n — APPROVED** [review]

**Pull request merged**